### PR TITLE
LPS-176146 Create and display FDS Entries

### PR DESCRIPTION
**Pull request by markocikos**
### References

- [LPS-176146 Create and display FDS Entries](https://issues.liferay.com/browse/LPS-176146)
- [Figma mockups](https://www.figma.com/file/VTOLTBFe99y9kA7AY3TcNR/lps-175276-design-app-to-manage-datasets?node-id=68%3A99382&t=S7PI9NDt5ONIuAzY-0)

### What is the goal of this PR?

In this PR, we are:
- adding persistence layer for FDS Entries
- creating UI to add and display FDS Entries
- extending API of FDS `CreationMenu`, required by solution
- extending API of `openModal`, to improve code quality of solution
- adding a minor fix for omitting unnecessary request param 

Notes:
- `FDS Entries` is our top-level entity, with attributes:
  - `label`, user input text
  - `entityClassName`. Unique identifier of headless resource. From this we can construct `Provider` in UI, containing all resource information.
  - We will implement Views count, seen in mockups,in a separate story 
- Despite starting with entries, I think it's still ok to call the module `frontend-data-set-views-web`. The main focus later on will be on FDS Views, multiple for each FDS Entry.
- There is a known bug that headless resource created by Objects API does not look right in selection. We will fix this in a separate task.
- We are using Objects API for backend! :tada: I believe this is the first "real" usage in DXP. We are following the same pattern of object creation as in FDS Sample module. cc @gabrielwas. The fact that we are working with async requests for CRUD actions pushed us to implement as much as possible in frontend only, without page reload. This was the main reason to creation from dedicated page to modal, to avoid navigation action.  
- extending API of FDS `CreationMenu` allows us to reload data in creation menu action workflow.
- I extending API of `openModal`, enable setting of entire modal content as one React component. This allows us to have good scoping of states, where validation is internal to modal. It is technically possible to work around it, like we did for [custom views](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/CustomViewsControls.js), but it leaves code harder to understand and with a bit worse performance.

cc @ugeortiz @manoelcyreno

### What does it look like?

https://user-images.githubusercontent.com/5435511/220976462-3a746c7e-ef94-4170-95c7-0ee71605f783.mp4

### Steps to reproduce

1. Add `feature.flag.LPS-164563=true` to `portal-ext.properties`
2. Open Global Menu > Control Panel > Object > Datasets

### CODE REVIEW

**Comment by eltonccdantas** [comment]
ci:bundle

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSEntries.tsx [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+	headlessResources: initialHeadlessResources,
+	setHeadlessResourceValidationError,
+	setSelectedHeadlessResource,
+}: IDropdownMenuProps) => {
+	const [headlessResources, setHeadlessResources] = useState<
+		Array<HeadlessResource>
+	>(initialHeadlessResources || []);
+	const [query, setQuery] = useState('');
+
+	const onSearch = (query: string) => {
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSEntries.tsx:118]: We load all headless resources at once from the backend. So we won't flood server with search queries on headless resources. Still, we should consider `throttle()` ing this and run it when `query.lenght>1`. Not needed for the PR to be fwd, just keep it in mind

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/modal/Modal.js [hunk 2]
```diff
@@ -156,6 +157,100 @@ const Modal = ({
...
+					'modal-body-iframe': url,
+				})}
+				style={{
+					height,
+				}}
+			>
+				{url && (
+					<>
+						{loading && <ClayLoadingIndicator />}
+						<Iframe
```
Comment (kresimir-coko) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/modal/Modal.js:187]: I believe we want an empty line between these 2.

**Review by kresimir-coko — CHANGES_REQUESTED** [review]
LGTM apart from a tiny SF comment

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSEntries.tsx [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+
+const FDSEntries = ({
+	apiURL,
+	headlessResources,
+	namespace,
+}: IFDSEntriesProps) => {
+	const headlessResourcesMapRef = useRef<Map<string, HeadlessResource>>();
+
+	useEffect(() => {
+		headlessResourcesMapRef.current = new Map(
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSEntries.tsx:180]: A minor nit, I _think_ this effect is not needed, we can just compute the ref in the beginning as nobody is changing the headlessResources array between 2 renders.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSEntries.tsx (hunk 1 above)
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSEntries.tsx:118]: I tried implementing debouncing, but there is a complication with TS and our `debounce` util. We [allow](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/index.d.ts#L74) only parameters of type `() => void`, while our search function has the search text argument. There may a workaround, there may be a good reason for this restriction, or I am maybe missing something here. I'd definitely leave this as a separate task. cc @kresimir-coko

Comment (kresimir-coko) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSEntries.tsx:118]: IIRC we used `() => void` as a catch-all type for functions without arguments. Not sure if there's a workaround that could fit our case. Maybe @bryceosterhaus has something in mind, seeing as you worked more with TS than me or Marko.

**Comment by markocikos** [comment]
Force-pushed! Changes:
- fixed conflicts in backend
- added commit: SF based on https://github.com/liferay-frontend/liferay-portal/pull/3072#discussion_r1119051042
- added commit: simplification based on https://github.com/liferay-frontend/liferay-portal/pull/3072#discussion_r1120583161

**Comment by markocikos** [comment]
ci:test:relevant

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSEntries.tsx (hunk 1 above)
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSEntries.tsx:118]: Yeah, the problem is that our `onSearch` function, that we want to debounce, is not without arguments.

**Review by kresimir-coko — APPROVED** [review]

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSEntries.tsx [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+		] = useState<HeadlessResource>();
+		const [labelValidationError, setLabelValidationError] = useState(false);
+		const [
+			headlessResourceValidationError,
+			setHeadlessResourceValidationError,
+		] = useState(false);
+
+		const fdsEntryLabelRef = useRef<HTMLInputElement>(null);
+
+		const addFDSEntry = async () => {
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSEntries.tsx:223]: not in this PR, we will need to abstract the persistence out of the components when more complex cases arise (e.g. more than 1 fetch per operation. It's more than fine like this for now

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/CreationMenu.js [hunk 5]
```diff
@@ -51,7 +53,9 @@ function CreationMenu({primaryItems}) {
 
 											setActive(false);
 
-											item.onClick?.();
+											item.onClick?.({
+												loadData,
+											});
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/CreationMenu.js:58]: I still can't believe we did not this before :man_shrugging: 
Very nice!

**Review by dsanz — APPROVED** [review]

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/portlet/FDSViewsPortlet.java [hunk 6]
```diff
@@ -79,13 +93,67 @@ protected void doDispatch(
...
+				Arrays.asList(
+					ObjectFieldUtil.createObjectField(
+						ObjectFieldConstants.BUSINESS_TYPE_TEXT,
+						ObjectFieldConstants.DB_TYPE_STRING, true, false, null,
+						_language.get(locale, "name"), "label", true),
+					ObjectFieldUtil.createObjectField(
+						ObjectFieldConstants.BUSINESS_TYPE_TEXT,
+						ObjectFieldConstants.DB_TYPE_STRING, true, false, null,
+						"entityClassName", "entityClassName", true)));
+
```
[the remark this answers is not in the document]
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/portlet/FDSViewsPortlet.java:143]: Yeah, I don't expect this to be final version. During development, I expect a lot of database droppings. For updates later on, we might need to introduce some form of versioning of `FDS Entry` objects.

**Comment by deborasoliveira** [comment]
Hi @markocikos! 
Please check CI stable routine so I can analyze the relevant results 👍🏽

**Comment by mateusralv** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
ci:forward

**Comment by markocikos** [comment]
CI failures are related. Working on it!

**Comment by markocikos** [comment]
Force-pushed! Changes:
- Simplified changes in `Modal.js`. Previous version had a bug that caused infinite refreshing of the component, causing all modals in DXP to fail.

**Comment by markocikos** [comment]
ci:forward

**Comment by deborasoliveira** [comment]
Hi @markocikos!
According to the [comparison](https://testray.liferay.com/home/-/testray/runs/compare?formDate=1677869207526&testrayRunIdA=2217658160&testrayRunIdB=2220374403), the failures seem unrelated to this PR and the stable routine failure is caused by Jenkins trying to generate a report.

cc: @mateusralv

**Comment by markocikos** [comment]
I'll try it once or twice more, manually forward after.

**Comment by markocikos** [comment]
ci:forward

**Comment by markocikos** [comment]
ci:forward

**Comment by markocikos** [comment]
Manual forward: https://github.com/brianchandotcom/liferay-portal/pull/131261

**Pull request closed without merging**